src/Stack/Script.hs

```diff
@@ -205,20 +208,34 @@ blacklist = Set.fromList
     , mkPackageName "cryptohash-sha256"
     ]
 
-toModuleInfo :: LoadedSnapshot -> ModuleInfo
-toModuleInfo ls =
-      mconcat
-    $ map (\(pn, lpi) ->
-            ModuleInfo
-            $ Map.fromList
-            $ map (, Set.singleton pn)
-            $ Set.toList
-            $ lpiExposedModules lpi)
-    $ filter (\(pn, lpi) ->
-            not (lpiHide lpi) &&
-            pn `Set.notMember` blacklist)
-    $ Map.toList
-    $ Map.union (void <$> lsPackages ls) (void <$> lsGlobals ls)
+getModuleInfo :: HasEnvConfig env => RIO env ModuleInfo
+getModuleInfo = do
+    sourceMap <- view $ envConfigL . to envConfigSourceMap
+    installMap <- toInstallMap sourceMap
+    (_installedMap, globalDumpPkgs, snapshotDumpPkgs, _localDumpPkgs) <-
+        getInstalled
+            GetInstalledOpts
+            { getInstalledProfiling = False
+            , getInstalledHaddock = False
+            , getInstalledSymbols = False
+            }
+            installMap
+    return $
+        toModuleInfo (notHidden $ smDeps sourceMap) snapshotDumpPkgs <>
+        toModuleInfo (smGlobal sourceMap) globalDumpPkgs
+  where
+    notHidden = Map.filter (not . dpHidden)
+    toModuleInfo pkgs dumpPkgs =
+        let pnames = Map.keysSet pkgs `Set.difference` blacklist
+            modules =
+                Map.fromList
```

Comment:
Correct me if I am wrong, but we can potentially have module names clash amongst different global packages, by converting to `Map` from list we will end up dropping information about those packages. My assumption is that it is exactly the reason for  `Set` in `ModuleInfo`. Consider this example:
* module `Data.Foo` comes from package `foo`
* module `Data.Foo` comes from package `bar`
We could still use both packages in our library by using `PackageImports`, but with approach employed here we will end up with one that is last in the list, eg. `("Data.Foo", Set.singleton "foo")` and info about `bar` is lost.

But I assume the desired result in above example would be  `Map.fromlIst [("Data.Foo", Set.fromList ["foo", "bar"])]`

Reply:
Yeah, you're exactly right and in the previous version it was done with `mconcat`, so I'll just use `Map.fromListWith mappend`